http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitorInterp.scala

```diff
@@ -0,0 +1,698 @@
+    implicit ev: ApplicativeAsk[F, AppContext]
+  ): F[CheckResult] =
+    for {
+      instance <- googleComputeService.getInstance(
+        runtimeAndRuntimeConfig.runtime.googleProject,
+        config.gceZoneName,
+        InstanceName(runtimeAndRuntimeConfig.runtime.runtimeName.asString)
+      )
+      result <- runtimeAndRuntimeConfig.runtime.status match {
+        case RuntimeStatus.Creating =>
+          creatingRuntime(instance, monitorContext, runtimeAndRuntimeConfig)
+        // This is needed because during boot time, we no longer have reference to the previous delete operation
+        // But this path should only happen during boot and there's on-going delete
+        case RuntimeStatus.Deleting =>
+          instance match {
+            case None => deletedRuntime(monitorContext, runtimeAndRuntimeConfig)
+            case _    => checkAgain(monitorContext, runtimeAndRuntimeConfig, Some("Instance hasn't been deleted yet"))
+          }
+        case RuntimeStatus.Starting =>
+          startingRuntime(instance, monitorContext, runtimeAndRuntimeConfig)
+        // This is needed because during boot time, we no longer have reference to the previous delete operation
+        // But this path should only happen during boot and there's on-going delete
+        case RuntimeStatus.Stopping =>
+          val gceStatus = instance.flatMap(i => GceInstanceStatus.withNameInsensitiveOption(i.getStatus))
+          gceStatus match {
+            case None =>
+              logger
+                .error(s"${monitorContext} | Can't stop an instance that hasn't been initialized yet or doesn't exist")
+                .as(((), None)) //TODO: Ideally we should have sentry report this case
+            case Some(s) =>
+              s match {
+                case Stopped | Terminated =>
+                  stopRuntime(runtimeAndRuntimeConfig, monitorContext)
+                case _ =>
+                  checkAgain(
+                    monitorContext,
+                    runtimeAndRuntimeConfig,
+                    Some(s"hasn't been terminated yet")
+                  )
+              }
+          }
+        case status => logger.error(s"${status} is not a transition status for GCE; hence no need to monitor").as(((), None))
+      }
+    } yield result
+
+  private[monitor] def creatingRuntime(
+    instance: Option[Instance],
+    monitorContext: MonitorContext,
+    runtimeAndRuntimeConfig: RuntimeAndRuntimeConfig
+  )(implicit ev: ApplicativeAsk[F, AppContext]): F[CheckResult] = instance match {
+    case None =>
+      checkAgain(monitorContext, runtimeAndRuntimeConfig, Some(s"Can't retrieve instance yet"))
+    case Some(i) =>
+      for {
+        gceStatus <- F.fromEither(GceInstanceStatus.withNameInsensitiveOption(i.getStatus).toRight(new Exception(s"Unknown GCE instance status ${i.getStatus}"))) //TODO: Ideally we should have sentry report this case
```

Comment:
If there is an unexpected status then the stream will log an error and just stop, right? Agreed about having alerting about this.

I think the old monitor would just restart if an error occurred. Which meant that we had some really old monitor actor instances continually logging errors. I think I like failing fast better (and alerting).

Reply:
Thinking out loud: should we map to `failedRuntime` in this case instead of throwing?

Reply:
failRuntime will delete the runtime. I'm thinking this case will happen if gce introduces a new status that we don't know about, so I don't think we should blindly delete the runtime

Reply:
oh for previous comment, if this errors, the stream will terminate

```
@ Stream.unfoldLoopEval[IO, String, String]("")(s => if(s.length < 10) IO.raiseError(new Exception("error")) else IO.pure((s, None)))
res6: Stream[IO, String] = Stream(..)

@ res6.compile.toList
res7: IO[List[String]] = Map(
  Bind(Delay(fs2.Stream$CompileOps$$Lambda$2686/1839813660@641c8ba4), fs2.Stream$Compiler$$anon$3$$Lambda$2689/506027903@52667676),
  fs2.Stream$CompileOps$$Lambda$2688/876029877@3fad8636,
  0
)

@ res7.unsafeRunSync
java.lang.Exception: error
  ammonite.$sess.cmd6$.$anonfun$res6$1(cmd6.sc:1)
  fs2.Stream$.$anonfun$unfoldLoopEval$1(Stream.scala:3669)
  fs2.Stream$.$anonfun$unf
```